File: www/update.php
<?php

require_once '/etc/polony-tools/config.php';
require_once 'functions.php';
set_time_limit(600);

mysql_connect($mogilefs_mysql_host,
	      $mogilefs_mysql_username,
	      $mogilefs_mysql_password);
echo mysql_error();

mysql_select_db($mogilefs_mysql_database);
echo mysql_error();

lock_or_exit("dataset");

putenv("MOGILEFS_TRACKERS=".join(",",$mogilefs_trackers));

function dbsetup()
{
  global $analysis_mysql_host,
    $analysis_mysql_username,
    $analysis_mysql_password,
    $analysis_mysql_database;
  mysql_connect($analysis_mysql_host,
		$analysis_mysql_username,
		$analysis_mysql_password);
  echo mysql_error();
  mysql_select_db($analysis_mysql_database);
  echo mysql_error();
  mysql_query("
create table
 if not exists
 dataset
(
 dsid char(32) not null primary key,
 nframes int,
 ncycles int
)");
  mysql_query("
create table
 if not exists
 cycle
(
 dsid char(32) not null,
 cid char(32) not null,
 nfiles int,
 nbytes bigint,
 exposure varchar(255),
 unique(dsid,cid)
)");
  echo mysql_error();
}

function grok ($fid=undef, $length=undef, $dmid=undef, $dkey=undef)
{
  global $grok;
  global $namespace;
  global $positions;
  global $cycles;
  if ($fid !== undef)
    {
      $dkey = explode("/", ereg_replace("^/*","",$dkey));
      $subdir = $dkey[count($dkey)-2];
      $dataset = $dkey[0];
      if (!isset($grok[$dataset]))
	{
	  $grok[$dataset] = array();
	}
      if (!isset($grok[$dataset][$subdir]))
	{
	  $grok[$dataset][$subdir] = array();
	  $grok[$dataset][$subdir][length] = 0;
	  $grok[$dataset][$subdir][count] = 0;
	}
      $grok[$dataset][$subdir][length] += $length;
      $grok[$dataset][$subdir][count] ++;
    }
  else
    {
      dbsetup();
      echo "Updating dataset and cycle tables.\n";
      mysql_query("lock tables dataset write, cycle write");
      mysql_query("delete from dataset");
      mysql_query("delete from cycle");
      foreach ($grok as $dataset => $d)
	{
	  putenv("MOGILEFS_DOMAIN=images");
	  $key = escapeshellarg("/".$dataset."/IMAGES/RAW/cycles");
	  $cycles = file_get_contents(`./moggetpaths $key`);
	  $key = escapeshellarg("/".$dataset."/IMAGES/RAW/positions");
	  $positions = file_get_contents(`./moggetpaths $key`);
	  $nframes = 0;
	  foreach (explode("\n", $positions) as $p)
	    {
	      if (ereg ("^[0-9]+[ \t]", $p))
		{
		  ++$nframes;
		}
	    }
	  $ncycles = 0;
	  $cycleinfo = array();
	  foreach (explode("\n", $cycles) as $c)
	    {
	      if ($c == "") continue;
	      $csv = explode(",", $c);
	      ++$ncycles;
	      mysql_query ("
		insert ignore into cycle
		(dsid, cid)
		values
		('$dataset', '$csv[1]')");
	      mysql_query ("
		update cycle
		set exposure='$c'
		where dsid='$dataset' and cid='$csv[1]'");
	    }
	  mysql_query ("
	    replace into dataset
	    (dsid, nframes, ncycles)
	    values
	    ('$dataset',
	     '$nframes',
	     '$ncycles')");
	  foreach ($d as $subdir => $s)
	    {
	      mysql_query("
		insert ignore into cycle
		(dsid, cid)
		values
		('$dataset', '$subdir')");
	      mysql_query("
		update cycle set
		nfiles='$s[count]',
		nbytes='$s[length]'
		where dsid='$dataset' and cid='$subdir'");
	    }
	}
      mysql_query("unlock tables");
      echo "Done.\n";
    }
}

$namespace = array();
$dmid_images = mysql_one_value("select dmid from domain where namespace='images'");

$q = mysql_query("select
     file.fid,
     file.length,
     file.dmid,
     file.dkey
     from file
     where dmid='$dmid_images'
     order by dmid,dkey");
echo mysql_error();
$lastfid = -1;
while($row = mysql_fetch_row($q))
{
  list ($fid, $length, $dmid, $dkey) = $row;
  if ($fid == $lastfid) continue;
  $lastfid = $fid;
  if (ereg("/IMAGES/RAW/positions$", $dkey)) continue;
  if (ereg("/IMAGES/RAW/cycles$", $dkey)) continue;
  grok($fid, $length, $dmid, $dkey);
}
grok();

// arch-tag: acc56792-0004-11dc-9207-0015f2b17887
?>

File: www/update.pl
#!/usr/bin/perl

use strict;
use MogileFS::Client;
use Digest::MD5 'md5_hex';
use DBI;
use Fcntl ':flock';

do '/etc/polony-tools/config.pl';


debuglog ("Connecting to mysql");


my $dbh = DBI->connect($main::analysis_dsn,
		       $main::mrwebgui_mysql_username,
		       $main::mrwebgui_mysql_password)
    or die DBI->errstr;


debuglog ("Looking up dmid for $main::mogilefs_default_domain");

my $sth = $dbh->prepare ("select dmid from mogilefs.domain where namespace=?");
$sth->execute ($main::mogilefs_default_domain) or die DBI->errstr;
my ($dmid) = $sth->fetchrow_array;
die "Couldn't find dmid for namespace '$main::mogilefs_default_domain'"
    unless $dmid > 0;


debuglog ("Locking");

open LOCKFILE, "+>>/var/lock/lock.polony-tools.update" or die "lockfile open";
flock LOCKFILE, LOCK_EX or die "lockfile flock";


debuglog ("Reading all MogileFS keys");

my %count;
my %size;
my $after = "";
my $lastdsid;
my $lastcycle;
my $lengthcache;
my $countcache;

$sth = $dbh->prepare ("select dkey,length
 from mogilefs.file
 where dmid=?
 and dkey>?
 order by dkey")
    or die DBI->errstr;

$sth->{"mysql_use_result"} = 1;	# do not try to buffer the result at client

while (defined $after)
{
    $sth->execute ($dmid, $after)
	or die DBI->errstr;
    $after = undef;
    while (my ($dkey, $length) = $sth->fetchrow_array)
    {
	$after = $dkey;
	my ($dsid, $cycle);
	if ($dkey =~ m,^/([^/]+)/IMAGES/RAW/([^/]+)/[^/]*\.(tiff?|raw)(\.g?z)?$,)
	{
	    $dsid = $1;
	    $cycle = $2;
	    $cycle =~ tr/A-Z/a-z/;
	}
	elsif ($dkey =~ m,^/([^/]+), && $dkey !~ m,^/\d+/frame/,)
	{
	    $dsid = $1;
	    $cycle = "none";
	}
	else
	{
	    $dsid = "none";
	    $cycle = "none";
	}
	if ($lastdsid ne $dsid || $lastcycle ne $cycle)
	{
	    if (defined $lastdsid)
	    {
		if (!ref $size{$lastdsid})
		{
		    $size{$lastdsid} = {};
		    $count{$lastdsid} = {};
		    print STDERR "$lastdsid\n";
		}
		$size{$lastdsid}{$lastcycle} += $lengthcache;
		$count{$lastdsid}{$lastcycle} += $countcache;
	    }
	    $lengthcache = 0;
	    $countcache = 0;
	}
	$lastdsid = $dsid;
	$lastcycle = $cycle;
	$lengthcache += $length;
	$countcache ++;
	print STDERR "." if ($countcache % 1000 == 0);
    }
}

if (defined $lastdsid)
{
    if (!ref $size{$lastdsid})
    {
	$size{$lastdsid} = {};
	$count{$lastdsid} = {};
	print STDERR "$lastdsid\n";
    }
    $size{$lastdsid}{$lastcycle} += $lengthcache;
    $count{$lastdsid}{$lastcycle} += $countcache;
    $lengthcache = 0;
    $countcache = 0;
}


debuglog ("Connecting to MogileFS");

my $mogc;
for (qw(10 20 30 40 50))
{
    $mogc = eval {
	MogileFS::Client->new (domain => $main::mogilefs_default_domain,
			       hosts => [@main::mogilefs_trackers]);
      };
    last if $mogc;
    debuglog ("Failed, wait $_ seconds");
    sleep $_;
}
die "$@" if !$mogc;


debuglog ("Building dataset_tmp and cycle_tmp");

$dbh->do ("create table dataset_tmp like dataset");
$dbh->do ("delete from dataset_tmp"); # maybe it already existed
$dbh->do ("create table cycle_tmp like cycle");
$dbh->do ("delete from cycle_tmp"); # maybe it already existed

foreach my $dsid (sort keys %size)
{
    my %cycle;
    my $nframes;

    if (my $positions = $mogc->get_file_data ("/$dsid/IMAGES/RAW/positions"))
    {
	foreach (split ("\n", $$positions))
	{
	    ++$nframes if /^\d+\s/;
	}
    }

    if (my $cycles = $mogc->get_file_data ("/$dsid/IMAGES/RAW/cycles"))
    {
	foreach (split ("\n", $$cycles))
	{
	    if (my @c = split (","))
	    {
		if (@c > 1)
		{
		    $c[1] =~ tr/A-Z/a-z/;
		    $cycle{$c[1]} = 1;
		    $dbh->do ("insert into cycle_tmp (dsid, cid, exposure) values (?, ?, ?)",
			      undef, $dsid, $c[1], $_);
		}
	    }
	}
    }

    foreach my $cycle (sort keys %{$size{$dsid}})
    {
	if (!exists $cycle{$cycle})
	{
	    $cycle{$cycle} = 1;
	    $dbh->do ("insert into cycle_tmp (dsid, cid) values (?, ?)",
		      undef, $dsid, $cycle);
	}
	$dbh->do ("update cycle_tmp set nfiles=?, nbytes=? where dsid=? and cid=?",
		  undef,
		  $count{$dsid}{$cycle},
		  $size{$dsid}{$cycle},
		  $dsid,
		  $cycle);
    }

    my $ncycles = scalar keys %cycle;
    $dbh->do ("insert into dataset_tmp (dsid,nframes,ncycles) values (?,?,?)",
	      undef, $dsid, $nframes, $ncycles);
}


debuglog ("Copying dataset_tmp and cycle_tmp to dataset and cycle tables");

$dbh->do ("lock tables dataset write, cycle write, dataset_tmp write, cycle_tmp write") or die DBI->errstr;
$dbh->do ("delete from dataset") or die DBI->errstr;
$dbh->do ("delete from cycle") or die DBI->errstr;
$dbh->do ("insert into dataset select * from dataset_tmp") or die DBI->errstr;
$dbh->do ("insert into cycle select * from cycle_tmp") or die DBI->errstr;
$dbh->do ("unlock tables") or die DBI->errstr;
$dbh->do ("drop table dataset_tmp") or die DBI->errstr;
$dbh->do ("drop table cycle_tmp") or die DBI->errstr;


debuglog ("Done");


sub debuglog
{
    print STDERR (scalar localtime, " ", @_, "\n");
}
